pkg/querier/distributor_queryable.go

```diff
@@ -38,37 +39,59 @@ type Distributor interface {
 	LabelValuesCardinality(ctx context.Context, labelNames []model.LabelName, matchers []*labels.Matcher) (uint64, *client.LabelValuesCardinalityResponse, error)
 }
 
-func newDistributorQueryable(distributor Distributor, iteratorFn chunkIteratorFunc, queryIngestersWithin time.Duration, logger log.Logger) QueryableWithFilter {
+func newDistributorQueryable(distributor Distributor, iteratorFn chunkIteratorFunc, cfgProvider distributorQueryableConfigProvider, logger log.Logger) QueryableWithFilter {
 	return distributorQueryable{
-		logger:               logger,
-		distributor:          distributor,
-		iteratorFn:           iteratorFn,
-		queryIngestersWithin: queryIngestersWithin,
+		logger:      logger,
+		distributor: distributor,
+		iteratorFn:  iteratorFn,
+		cfgProvider: cfgProvider,
 	}
 }
 
+type distributorQueryableConfigProvider interface {
+	QueryIngestersWithin(userID string) time.Duration
+}
+
 type distributorQueryable struct {
-	logger               log.Logger
-	distributor          Distributor
-	iteratorFn           chunkIteratorFunc
-	queryIngestersWithin time.Duration
+	logger      log.Logger
+	distributor Distributor
+	iteratorFn  chunkIteratorFunc
+	cfgProvider distributorQueryableConfigProvider
 }
 
 func (d distributorQueryable) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	queryIngestersWithin := d.cfgProvider.QueryIngestersWithin(userID)
+	now := time.Now()
+
+	// queryIngestersWithin might have changed since d.UseQueryable() was called, so we check it again when creating a querier
+	if !d.useQueryable(now, mint, maxt, queryIngestersWithin) {
+		return storage.NoopQuerier(), nil
+	}
```

Comment:
You can remove this check. There's no practical value doing it. We get a Queryable for every query.

Reply:
It solves possible race: `queryIngestersWithin` may be different from when `d.UseQueryable()` was called, and now queryable is no longer needed. Rather than querying ingesters, we return empty querier.

> We get a Queryable for every query.

There is only one queryable, which returns new Querier for each query.

Reply:
I meant the `storage.QueryableFunc()` which is called for each query.

Reply:
> It solves possible race: queryIngestersWithin may be different from when d.UseQueryable() was called, and now querier is no longer needed. Rather than querying ingesters, we return empty querier.

How long is the time window of this race condition?

Reply:
Discussed on Slack with @pstibrany. Now I see the race condition. We discussed options to remove the double check and only have 1. He will follow up with a suggestion.

Reply:
> We discussed options to remove the double check and only have 1.

The idea is simple: we can make `distributorQueryable.UseQueryable` to always return `true` and only have a check in `distributorQueryable.Querier`.

As discussed privately, we may want to simplify this code a bit as well... get rid of `UseQueryable` and move the checks into `Querier` methods, but that's a topic for another PR.

Reply:
Updated in https://github.com/grafana/mimir/pull/4287/commits/45f624c5f17b8caa275bb8cf5e01f7133c7673d5